dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1540 @@
+                return winrt::TeachingTipPlacementMode::Right;
+            }
+            else if (leftCenterAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::Left;
+            }
+            else if (topRightAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::TopEdgeAlignedRight;
+            }
+            else if (topLeftAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::TopEdgeAlignedLeft;
+            }
+            else if (bottomRightAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::BottomEdgeAlignedRight;
+            }
+            else if (bottomLeftAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::BottomEdgeAlignedLeft;
+            }
+            else if (rightTopAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::RightEdgeAlignedTop;
+            }
+            else if (rightBottomAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::RightEdgeAlignedBottom;
+            }
+            else if (leftTopAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::LeftEdgeAlignedTop;
+            }
+            else if (leftBottomAvailable)
+            {
+                return winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom;
+            }
+        }
+    }
+    //the teaching tip wont fit anywhere, just return top.
+    return winrt::TeachingTipPlacementMode::Top;
+}
+
+
+void TeachingTip::EstablishShadows()
+{
+#ifdef USE_INSIDER_SDK
+    if (SharedHelpers::IsThemeShadowAvailable())
+    {
+#ifdef _DEBUG
+        auto beakShadow = winrt::Windows::UI::Xaml::Media::ThemeShadow{};
+        beakShadow.Receivers().Append(m_target.get());
+        m_beakPolygon.get().Shadow(beakShadow);
+        m_beakPolygon.get().Translation(winrt::float3{ m_beakPolygon.get().Translation().x, m_beakPolygon.get().Translation().y, m_beakElevation });
+#endif
+
+        auto contentShadow = winrt::Windows::UI::Xaml::Media::ThemeShadow{};
+        contentShadow.Receivers().Append(m_shadowTarget.get());
+        m_rootGrid.get().Shadow(contentShadow);
+        m_rootGrid.get().Translation(winrt::float3{ m_rootGrid.get().Translation().x, m_rootGrid.get().Translation().y, m_contentElevation });
+    }
+#endif
+}
+
+void TeachingTip::OnPropertyChanged(
+    const winrt::DependencyObject& sender,
+    const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    if (args.Property() == s_AttachProperty)
+    {
+        OnAttachPropertyChanged(sender, args);
+    }
+    else
+    {
+        winrt::get_self<TeachingTip>(sender.as<winrt::TeachingTip>())->OnPropertyChanged(args);
+    }
+}
+
+void TeachingTip::OnAttachPropertyChanged(
+    const winrt::DependencyObject& sender,
+    const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    auto oldTip = unbox_value<winrt::TeachingTip>(args.OldValue());
```

Comment:
I'm surprised unbox_value works -- args.OldValue().try_as<winrt::TeachingTip>() is what I would have expected since it's a reference type and the target might not be a TeachingTip.

Reply:
Since AttachProperty is an AttachedProperty, doesn't that mean the sender is a reference type which could be anything, but the OldValue and NewValue are going to be TeachingTips?

---
In reply to: [239325911](https://github.com/Microsoft/microsoft-ui-xaml/pull/32#discussion_r239325911) [](ancestors = 239325911)